akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -193,173 +274,74 @@ private[akka] class AffinityPool(
-          }
-        case Yielding ⇒
-          yields += 1
-          if (yields > maxYields) {
-            state = Parking
-            parkPeriodNs = minParkPeriodNs
-          } else Thread.`yield`()
-        case Parking ⇒
-          LockSupport.parkNanos(parkPeriodNs)
-          parkPeriodNs = Math.min(parkPeriodNs << 1, maxParkPeriodNs)
+  private final class AffinityPoolWorker( final val q: BoundedAffinityTaskQueue, final val idleStrategy: IdleStrategy) extends Runnable {
+    final val thread: Thread = tf.newThread(this)
+    @volatile private[this] var executing: Boolean = false
+
+    def startWorker(): Unit = thread.start()
+
+    override final def run(): Unit = {
+
+      def executeNext(): Unit = {
+        val c = q.poll()
+        if (c ne null) {
+          executing = true
+          try
+            c.run()
+          finally
+            executing = false
+          idleStrategy.reset()
+        } else {
+          idleStrategy.idle() // if not wait for a bit
+        }
       }
-    }
-
-    def reset(): Unit = {
-      spins = 0
-      yields = 0
-      state = NotIdle
-    }
-
-  }
-
-  private final class BoundedTaskQueue(capacity: Int) extends AbstractBoundedNodeQueue[Runnable](capacity)
-
-  private final class ThreadPoolWorker(val q: BoundedTaskQueue, val idleStrategy: IdleStrategy) extends Runnable {
-
-    private sealed trait WorkerState
-    private case object NotStarted extends WorkerState
-    private case object InExecution extends WorkerState
-    private case object Idle extends WorkerState
-
-    val thread: Thread = tf.newThread(this)
-    @volatile private var workerState: WorkerState = NotStarted
-
-    def startWorker(): Unit = {
-      workerState = Idle
-      thread.start()
-    }
-
-    private def runCommand(command: Runnable) = {
-      workerState = InExecution
-      try
-        command.run()
-      finally
-        workerState = Idle
-    }
-
-    override def run(): Unit = {
 
       /**
-       * Determines whether the worker can keep running or not.
-       * In order to continue polling for tasks three conditions
-       * need to be satisfied:
-       *
-       * 1) pool state is less than Shutting down or queue
-       * is not empty (e.g pool state is ShuttingDown but there are still messages to process)
-       *
-       * 2) the thread backing up this worker has not been interrupted
-       *
-       * 3) We are not in ShutDown state (in which we should not be processing any enqueued tasks)
+       * We keep running as long as we are Running
+       * or we're ShuttingDown but we still have tasks to execute,
+       * and we're not interrupted.
        */
-      def shouldKeepRunning =
-        (poolState < ShuttingDown || !q.isEmpty) &&
-          !Thread.interrupted() &&
-          poolState != ShutDown
+      @tailrec def runLoop(): Unit =
+        if (!Thread.interrupted()) {
+          (poolState: @switch) match {
+            case Uninitialized => ()
+            case Running ⇒
```

Comment:
@zaharidichev I introduced an Uninitialized state, both to make the different states clearer, but also to make Running the dead-center (5 >>> 1 == 2) of a possible binary search instead of a jmp-table for the table/lookupswitch bytecode.